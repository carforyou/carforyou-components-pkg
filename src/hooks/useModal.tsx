import { createPortal } from "react-dom"
import React, { RefObject, useCallback, useEffect, useState } from "react"

import classNames from "classnames"

import { disableScrollOnBody, enableScrollOnBody } from "../lib/scrollHelper"
import Modal, {
  ModalOverflow,
  ModalSize,
  ModalStyle,
} from "../components/modal/index"

export interface UseModalOptions {
  size?: ModalSize
  style?: ModalStyle
  alwaysRender?: boolean
  container?: RefObject<HTMLDivElement>
  verticalOverflow?: ModalOverflow
  onClose?: () => void
  title?: string
}

const useModal = (
  modal: (modalProps: { closeModal: () => void }) => JSX.Element,
  options?: UseModalOptions
) => {
  const [isVisible, setVisible] = useState(false)
  const {
    size = "medium",
    style = "white",
    alwaysRender = false,
    container = null,
    verticalOverflow = null,
    onClose = null,
    title = null,
  } = options || {}

  const openModal = useCallback(() => {
    disableScrollOnBody()
    setVisible(true)
  }, [])

  const closeModal = useCallback(() => {
    enableScrollOnBody()
    setVisible(false)
  }, [])

  useEffect(() => {
    return enableScrollOnBody
  }, [])

  const renderModalComponent = () => (
    <Modal
      close={closeModal}
      onClose={onClose}
      size={size}
      style={style}
      verticalOverflow={verticalOverflow}
      title={title}
    >
      {modal}
    </Modal>
  )

  const renderModalWrapper = () =>
    alwaysRender ? (
      <div className={classNames({ hidden: !isVisible })}>
        {renderModalComponent()}
      </div>
    ) : isVisible ? (
      renderModalComponent()
    ) : null

  const renderModal = () =>
    container?.current
      ? createPortal(renderModalWrapper(), container.current)
      : renderModalWrapper()

  return {
    openModal,
    closeModal,
    renderModal,
    modalVisible: isVisible,
  }
}

export default useModal
