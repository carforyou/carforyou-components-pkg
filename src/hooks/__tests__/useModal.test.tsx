import React, { createRef, FC, RefObject } from "react"
import { fireEvent, render, screen } from "@testing-library/react"

import useModal from "../useModal"

const modalText = "HERE BE DRAGONS"
const buttonText = "Open modal"

const Component: FC<{
  modalContainer?: RefObject<HTMLDivElement>
  onClose?: () => void
}> = ({ modalContainer, onClose = null }) => {
  const { renderModal, openModal } = useModal(() => <div>{modalText}</div>, {
    container: modalContainer,
    onClose,
  })

  return (
    <div>
      <button onClick={openModal}>{buttonText}</button>
      {renderModal()}
    </div>
  )
}

const ComponentWithPortal: FC = () => {
  const portalRef: RefObject<HTMLDivElement> = createRef()

  return (
    <>
      <div data-testid="modal-wrapper" ref={portalRef} />
      <Component modalContainer={portalRef} />
    </>
  )
}

describe("useModal", () => {
  describe("without portal", () => {
    const renderWrapper = (onClose = null) =>
      render(<Component onClose={onClose} />)

    it("renders correct markup", () => {
      const { container } = renderWrapper()

      expect(container.firstChild).toMatchSnapshot()
    })

    it("opens modal when clicking the button", () => {
      renderWrapper()

      fireEvent.click(screen.getByText(buttonText))

      expect(screen.getByText(modalText)).toMatchSnapshot()
    })

    it("renders the modal within the container", () => {
      const { container } = renderWrapper()

      fireEvent.click(screen.getByText(buttonText))

      expect(container).toMatchSnapshot()
    })

    it("closes the modal when pressing ESC", async () => {
      const { container } = renderWrapper()

      fireEvent.click(screen.getByText(buttonText))
      screen.getByText(modalText)
      fireEvent.keyDown(container, { keyCode: 27 })

      expect(container.firstChild).toMatchSnapshot()
      expect(container).not.toContain(modalText)
    })

    it("calls the onClose function when pressing ESC", async () => {
      const onCloseMock = jest.fn()
      const { container } = renderWrapper(onCloseMock)

      fireEvent.click(screen.getByText(buttonText))
      screen.getByText(modalText)
      fireEvent.keyDown(container, {
        keyCode: 27,
      })

      expect(onCloseMock).toBeCalledTimes(1)
    })

    it("calls the onClose function the close button is clicked", async () => {
      const onCloseMock = jest.fn()
      renderWrapper(onCloseMock)

      fireEvent.click(screen.getByText(buttonText))
      screen.getByText(modalText)
      fireEvent.click(screen.getByTestId("modal-close"))

      expect(onCloseMock).toBeCalledTimes(1)
    })

    it("calls the onClose function when the overlay is clicked", async () => {
      const onCloseMock = jest.fn()
      renderWrapper(onCloseMock)
      fireEvent.click(screen.getByText(buttonText))
      screen.getByText(modalText)
      fireEvent.click(screen.getByTestId("modal-overlay"))

      expect(onCloseMock).toBeCalledTimes(1)
    })
  })

  describe("with a portal", () => {
    const renderWrapper = () => render(<ComponentWithPortal />)

    it("renders the modal within the portal", () => {
      renderWrapper()

      fireEvent.click(screen.getByText(buttonText))

      expect(screen.getByTestId("modal-wrapper")).toMatchSnapshot()
    })
  })
})
