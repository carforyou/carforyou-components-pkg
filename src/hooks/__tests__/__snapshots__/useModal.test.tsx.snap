// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useModal with a portal renders the modal within the portal 1`] = `
<div
  data-testid="modal-wrapper"
>
  <div
    class="fixed inset-0 overflow-y-scroll scrolling-touch z-modal min-h-screen transition duration-200 flex flex-col justify-center items-center md:overflow-y-auto"
  >
    <div
      class="absolute inset-0 bg-grey-dark opacity-40 transition duration-200"
      data-testid="modal-overlay"
    />
    <div
      class="max-h-full scrolling-touch md:overflow-y-auto overflow-y-scroll fixed inline-block md:relative inset-x-0 md:align-middle md:inset-auto"
      role="dialog"
    >
      <div
        class="absolute z-modalClose cursor-pointer right-modalClose top-modalClose"
        data-testid="modal-close"
      >
        <svg
          height="48"
          viewBox="0 0 24 24"
          width="48"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            fill="none"
            fill-rule="evenodd"
            stroke="#232A36"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.2"
          >
            <path
              d="M7 7l10 10M17 7L7 17"
            />
          </g>
        </svg>
      </div>
      <div
        class="z-modal w-12/12 my-0 rounded-4 bg-white p-40 md:w-modal"
      >
        <div>
          HERE BE DRAGONS
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`useModal without portal closes the modal when pressing ESC 1`] = `
<div>
  <button>
    Open modal
  </button>
</div>
`;

exports[`useModal without portal opens modal when clicking the button 1`] = `
<div>
  HERE BE DRAGONS
</div>
`;

exports[`useModal without portal renders correct markup 1`] = `
<div>
  <button>
    Open modal
  </button>
</div>
`;

exports[`useModal without portal renders the modal within the container 1`] = `
<div>
  <div>
    <button>
      Open modal
    </button>
    <div
      class="fixed inset-0 overflow-y-scroll scrolling-touch z-modal min-h-screen transition duration-200 flex flex-col justify-center items-center md:overflow-y-auto"
    >
      <div
        class="absolute inset-0 bg-grey-dark opacity-40 transition duration-200"
        data-testid="modal-overlay"
      />
      <div
        class="max-h-full scrolling-touch md:overflow-y-auto overflow-y-scroll fixed inline-block md:relative inset-x-0 md:align-middle md:inset-auto"
        role="dialog"
      >
        <div
          class="absolute z-modalClose cursor-pointer right-modalClose top-modalClose"
          data-testid="modal-close"
        >
          <svg
            height="48"
            viewBox="0 0 24 24"
            width="48"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
              stroke="#232A36"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.2"
            >
              <path
                d="M7 7l10 10M17 7L7 17"
              />
            </g>
          </svg>
        </div>
        <div
          class="z-modal w-12/12 my-0 rounded-4 bg-white p-40 md:w-modal"
        >
          <div>
            HERE BE DRAGONS
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
